docs/spec_draft.md

```diff
@@ -846,6 +847,151 @@ defined and one of the following:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.gather
+
+### Semantics
+
+Gathers slices from `operand` tensor from offsets specified in `start_indices`
+and produces a `result` tensor.
+
+The following diagram shows how elements in `result` map on elements in
+`operand` using a concrete example. The diagram picks a few example `result`
+indices and explains in detail which `operand` indices they correspond to.
+
+<img align="center" src="spec_draft/gather.png" />
+
+More formally, `result[result_index] = operand[operand_index]` where:
+  * `batch_dims` = [`d` for `d` in `axes(result)` and `d` not in `offset_dims`].
+  * `batch_index` = [`result_index[d]` for `d` in `batch_dims`].
+  * `start_index` =
+      * `start_indices[bi0, ..., :, ..., biN]` where `bi` are individual
+        elements in `batch_index` and `:` is inserted at the `index_vector_dim`
+        index, if `index_vector_dim` < `rank(start_indices)`.
+      * `[start_indices[batch_index]]` otherwise.
+  * For `do` in `axes(operand)`,
+      * `full_start_index[do]` = `start_index[ds]` if `do = start_index_map[ds]`.
+      * `full_start_index[do]` = `0` otherwise.
+  * `offset_index` = [`result_index[d]` for `d` in `offset_dims`].
+  * `full_offset_index` = `[oi0, ..., 0, ..., oiN]` where `oi` are individual
+    elements in `offset_index`, and `0` is inserted at indices from
+    `collapsed_slice_dims`.
+  * `operand_index` = `add(full_start_index, full_offset_index)`.
+    If `operand_index` is out of bounds for `operand`, then the behavior is
+    implementation-defined.
+
+If `indices_are_sorted` is `true` then the implementation can assume that the
+user sorts the `operand_index`s in the same order the corresponding tensor
+elements are laid out in memory. The latter is decided by the `start_index_map`.
+If they are not, then the behavior is implementation-defined.
```

Comment:
I got to say that I'm still confused by this description. E.g. the "Types" section says "This specification does not define in which order tensor elements are laid out in memory", so referring to this order doesn't add clarity. I think a formal notation would help to clarify things.

Reply:
formalized it. PTAL.